plugins/modem-manager/fu-mm-device.c

```diff
@@ -917,6 +917,135 @@ fu_mm_device_detach_fastboot(FuDevice *device, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_mm_device_mbim_open(FuMmDevice *self, GError **error)
+{
+	self->mbim_qdu_updater = fu_mbim_qdu_updater_new(self->port_mbim);
+	return fu_mbim_qdu_updater_open(self->mbim_qdu_updater, error);
+}
+
+static gboolean
+fu_mm_device_mbim_close(FuMmDevice *self, GError **error)
+{
+	g_autoptr(FuMbimQduUpdater) updater = NULL;
+	updater = g_steal_pointer(&self->mbim_qdu_updater);
+	return fu_mbim_qdu_updater_close(updater, error);
+}
+
+#if MBIM_CHECK_VERSION(1, 27, 5)
+static void
+fu_mm_device_switch_to_edl_mbim_ready(MbimDevice *device, GAsyncResult *res, GMainLoop *loop)
+{
+	/* No need to check for a response since MBIM
+	 * port goes away without sending one */
+
+	g_main_loop_quit(loop);
+}
+
+static gboolean
+fu_mm_device_mbim_switch_to_edl(FuDevice *device, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(MbimMessage) message = NULL;
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new(NULL, FALSE);
+
+	locker = fu_device_locker_new_full(device,
+					   (FuDeviceLockerFunc)fu_mm_device_mbim_open,
+					   (FuDeviceLockerFunc)fu_mm_device_mbim_close,
+					   error);
+	if (locker == NULL)
+		return FALSE;
+
+	message = mbim_message_qdu_quectel_reboot_set_new(MBIM_QDU_QUECTEL_REBOOT_TYPE_EDL, NULL);
+	mbim_device_command(fu_mbim_qdu_updater_get_mbim_device(self->mbim_qdu_updater),
+			    message,
+			    5,
+			    NULL,
+			    (GAsyncReadyCallback)fu_mm_device_switch_to_edl_mbim_ready,
+			    mainloop);
+
+	g_main_loop_run(mainloop);
+
+	return TRUE;
+}
+#endif // MBIM_CHECK_VERSION(1, 27, 5)
+
+static gboolean
+fu_mm_device_qcdm_switch_to_edl_cb(FuDevice *device, gpointer userdata, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	static const guint8 emergency_download[] = {0x4b, 0x65, 0x01, 0x00, 0x54, 0x0f, 0x7e};
+
+	/* when the QCDM port does not exist anymore, we are already detached */
+	if (!g_file_test(self->port_qcdm, G_FILE_TEST_EXISTS))
+		return TRUE;
+
+	locker = fu_device_locker_new_full(self,
+					   (FuDeviceLockerFunc)fu_mm_device_io_open_qcdm,
+					   (FuDeviceLockerFunc)fu_mm_device_io_close,
+					   error);
+	if (locker == NULL)
+		return FALSE;
+
+	return fu_mm_device_qcdm_cmd(self,
+				     emergency_download,
+				     G_N_ELEMENTS(emergency_download),
```

Comment:
i think `sizeof()` is more appropriate here.

Reply:
I think we can even move the sizeof into the function itself, since `sizeof` is always applied to the parameter anyways.